src/datanode/src/sql/copy_table.rs

```diff
@@ -0,0 +1,147 @@
+
+        let accessor = Builder::default().build().unwrap();
+        let object_store = ObjectStore::new(accessor);
+
+        let mut parquet_writer = ParquetWriter::new(req.file_name, stream, object_store);
+        // TODO(jiachun):
+        // For now, COPY is implemented synchronously.
+        // When copying large table, it will be blocked for a long time.
+        // Maybe we should make "copy" runs in background?
+        // Like PG: https://www.postgresql.org/docs/current/sql-copy.html
+        let rows = parquet_writer.flush().await?;
+
+        Ok(Output::AffectedRows(rows))
+    }
+}
+
+type DfRecordBatchStream = Pin<Box<DfRecordBatchStreamAdapter>>;
+
+struct ParquetWriter {
+    file_name: String,
+    stream: DfRecordBatchStream,
+    object_store: ObjectStore,
+    max_row_group_size: usize,
+    max_rows_in_segment: usize,
+}
+
+impl ParquetWriter {
+    pub fn new(file_name: String, stream: DfRecordBatchStream, object_store: ObjectStore) -> Self {
+        Self {
+            file_name,
+            stream,
+            object_store,
+            // TODO(jiachun): make these configurable
+            max_row_group_size: 4096,
+            max_rows_in_segment: 1000000,
+        }
+    }
+
+    pub async fn flush(&mut self) -> Result<usize> {
+        let schema = self.stream.as_ref().schema();
+        let writer_props = WriterProperties::builder()
+            .set_compression(Compression::ZSTD)
+            .set_encoding(Encoding::PLAIN)
+            .set_max_row_group_size(self.max_row_group_size)
+            .build();
+        let mut total_rows = 0;
+        loop {
+            let mut buf = vec![];
+            let mut arrow_writer =
+                ArrowWriter::try_new(&mut buf, schema.clone(), Some(writer_props.clone()))
+                    .context(error::WriteParquetSnafu)?;
+
+            let mut rows = 0;
+            let mut end_loop = true;
+            // TODO(hl & jiachun): Since OpenDAL's writer is async and ArrowWriter requires a `std::io::Write`,
+            // here we use a Vec<u8> to buffer all parquet bytes in memory and write to object store
+            // at a time. Maybe we should find a better way to brige ArrowWriter and OpenDAL's object.
+            while let Some(batch) = self
+                .stream
+                .try_next()
+                .await
+                .context(error::PollStreamSnafu)?
+            {
+                arrow_writer
+                    .write(&batch)
+                    .context(error::WriteParquetSnafu)?;
+                rows += batch.num_rows();
+                if rows >= self.max_rows_in_segment {
+                    end_loop = false;
+                    break;
+                }
+            }
+
+            let start_row_num = total_rows + 1;
+            total_rows += rows;
+            arrow_writer.close().context(error::WriteParquetSnafu)?;
+
+            // if rows == 0, we just end up with a empty file.
```

Comment:
Can we just skip write if rows == 0?

Reply:
Originally I thought the same as you, and for this reason @MichaelScofield has offered a different opinion, which I think also has some merit:

"Better not early return here, because when copying empty table, we could end up with an empty file (instead of not existing), which is more user friendly. So I think we can remove this "if"."